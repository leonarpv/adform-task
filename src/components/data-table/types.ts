import { ReactNode } from 'react'

export type ColumnDefinition<TSource extends {}> = {
    accessor: (source: TSource) => ReactNode
    title: string
}
