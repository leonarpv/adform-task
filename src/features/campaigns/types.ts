export type Campaign = {
    id: number
    name: string
    startDate: string
    endDate: string
    Budget: number
    userId: number
}

export type CampaignUser = {
    id: number
    name: string
}

export type CampaignWithUser = {
    user: CampaignUser | null
} & Campaign

export type CampaignFilters = {
    name?: string
    startDate?: string
    endDate?: string
}
