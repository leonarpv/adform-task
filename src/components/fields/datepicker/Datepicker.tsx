import React from 'react'
import DatePicker, { ReactDatePickerProps } from 'react-datepicker'
import BaseField, { BaseFieldProps } from '../BaseField'
import './datepicker.scss'

type Props = ReactDatePickerProps & BaseFieldProps

const Datepicker = ({ label, error, ...rest }: Props) => {
    return (
        <BaseField label={label} error={error}>
            <DatePicker className="datepicker" {...rest} />
        </BaseField>
    )
}

export default Datepicker
