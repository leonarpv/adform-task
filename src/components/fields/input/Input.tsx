import React from 'react'
import BaseField, { BaseFieldProps } from '../BaseField'
import './input.scss'

type Props = React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
> &
    BaseFieldProps

const Input = ({ label, error, ...rest }: Props) => {
    return (
        <BaseField label={label} error={error}>
            <input className="input" {...rest} />
        </BaseField>
    )
}

export default Input
