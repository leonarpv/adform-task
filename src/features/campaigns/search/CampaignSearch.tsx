import React, { ChangeEvent, ReactNode } from 'react'
import { CampaignFilters } from '../types'
import Datepicker from 'components/fields/datepicker/Datepicker'
import Input from 'components/fields/input/Input'
import dayjs from 'dayjs'
import './campaignSearch.scss'

type Props = {
    children?: ReactNode
    filters?: CampaignFilters
    onChange: (search: CampaignFilters) => void
}

const CampaignSearch = ({ children, filters, onChange }: Props) => {
    const handleStartDateChange = (date: Date | null) => {
        onChange({ ...filters, startDate: date?.toString() })
    }

    const handleEndDateChange = (date: Date | null) => {
        onChange({ ...filters, endDate: date?.toString() })
    }

    const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
        onChange({ ...filters, name: e.target.value })
    }

    const startDate = filters?.startDate
        ? dayjs(filters.startDate).toDate()
        : undefined

    const endDate = filters?.endDate
        ? dayjs(filters.endDate).toDate()
        : undefined

    const maxStartDate = filters?.endDate
        ? dayjs(filters.endDate).subtract(1, 'day').toDate()
        : undefined

    const minEndDate = filters?.startDate
        ? dayjs(filters.startDate).add(1, 'day').toDate()
        : undefined

    return (
        <div className="campaign-search">
            <div className="campaign-search__inputs">
                <div className="campaign-search__date-pickers">
                    {children}
                    <Datepicker
                        label="Start Date"
                        isClearable
                        maxDate={maxStartDate}
                        selected={startDate}
                        onChange={handleStartDateChange}
                    />
                    <Datepicker
                        label="End Date"
                        isClearable
                        minDate={minEndDate}
                        selected={endDate}
                        onChange={handleEndDateChange}
                    />
                </div>

                <Input
                    label="Search By Name"
                    value={filters?.name || ''}
                    type="text"
                    onChange={handleNameChange}
                />
            </div>
        </div>
    )
}

export default CampaignSearch
