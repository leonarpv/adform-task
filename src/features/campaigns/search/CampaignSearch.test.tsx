import { shallow } from 'enzyme'
import React, { ChangeEvent, SyntheticEvent } from 'react'
import CampaignSearch from './CampaignSearch'
import { CampaignFilters } from '../types'
import Datepicker from 'components/fields/datepicker/Datepicker'
import Input from 'components/fields/input/Input'
import { Mock } from 'ts-mockery'

describe('CampaignSearch', () => {
    it('should render properly', () => {
        const filters: CampaignFilters = {}
        const onChange = jest.fn()

        const wrapper = shallow(
            <CampaignSearch filters={filters} onChange={onChange} />
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('should render datepickers and input', () => {
        const filters: CampaignFilters = {}
        const onChange = jest.fn()

        const wrapper = shallow(
            <CampaignSearch filters={filters} onChange={onChange} />
        )

        expect(wrapper.find(Datepicker).length).toBe(2)
        expect(wrapper.find(Input).exists()).toBe(true)
    })

    it('should call onChange on startDate datepicker change', () => {
        const filters: CampaignFilters = {}
        const onChange = jest.fn()

        const wrapper = shallow(
            <CampaignSearch filters={filters} onChange={onChange} />
        )

        const changedDate = new Date()
        const event = Mock.of<SyntheticEvent<HTMLInputElement, Event>>()

        wrapper.find(Datepicker).at(0).props().onChange(changedDate, event)

        expect(onChange).toHaveBeenCalledTimes(1)
        expect(onChange).toHaveBeenCalledWith({
            startDate: changedDate.toString(),
        })
    })

    it('should call onChange on endDate datepicker change', () => {
        const filters: CampaignFilters = {}
        const onChange = jest.fn()

        const wrapper = shallow(
            <CampaignSearch filters={filters} onChange={onChange} />
        )

        const changedDate = new Date()
        const event = Mock.of<SyntheticEvent<HTMLInputElement, Event>>()

        wrapper.find(Datepicker).at(1).props().onChange(changedDate, event)

        expect(onChange).toHaveBeenCalledTimes(1)
        expect(onChange).toHaveBeenCalledWith({
            endDate: changedDate.toString(),
        })
    })

    it('should call onChange on name input change', () => {
        const filters: CampaignFilters = {}
        const onChange = jest.fn()

        const wrapper = shallow(
            <CampaignSearch filters={filters} onChange={onChange} />
        )

        const event = Mock.of<ChangeEvent<HTMLInputElement>>({
            target: {
                value: 'value',
            },
        })

        const onInputChange = wrapper.find(Input).props().onChange

        expect(onInputChange).toBeDefined()

        onInputChange && onInputChange(event)

        expect(onChange).toHaveBeenCalledTimes(1)
        expect(onChange).toHaveBeenCalledWith({ name: event.target.value })
    })
})
