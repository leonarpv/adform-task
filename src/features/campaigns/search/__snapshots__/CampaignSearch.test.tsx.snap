// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignSearch should render properly 1`] = `ShallowWrapper {}`;
